### PR TITLE
Cache: slight change in naming

### PR DESCRIPTION
**Pull request by zucchini-nlp**
# What does this PR do?

Following https://github.com/huggingface/transformers/pull/32315#discussion_r1701676560, this PR adds two attributes on all cache classes, `is_sliding` and `is_static`. We can now rely on these attr to crop/prepare attention mask.

### CODE REVIEW

File: src/transformers/models/mistral/modeling_mistral.py [hunk 1]
```diff
@@ -891,11 +890,8 @@ def _update_causal_mask(
...
         min_dtype = torch.finfo(dtype).min
         sequence_length = input_tensor.shape[1]
-        # SlidingWindowCache
-        if using_sliding_window_cache:
-            target_length = max(sequence_length, self.config.sliding_window)
-        # StaticCache
-        elif using_static_cache:
+        # SlidingWindowCache or StaticCache
+        if past_key_values is not None and (past_key_values.is_sliding or past_key_values.is_static):
             target_length = past_key_values.get_max_length()
```
Comment (gante) [inline src/transformers/models/mistral/modeling_mistral.py:895]: Because of the changes on the sliding window cache, `target_length` can now be `None`, which means that `causal_mask = torch.full((sequence_length, target_length), ...)` will fail, a few lines below -- right?

I agree that the sliding window has infinite length. To untangle this, I have the following suggestion:
1. Let's rename `get_max_length` to `get_cache_shape` or a similar name. The idea is to have a function to retrieve the shape of the cache object, which is used in other places like here. (`get_max_length`  would have to go through a deprecation cycle, calling `get_cache_shape` under the hood until it is deleted )
2. Explore the current uses of `get_max_length()`. When we want to obtain the shape of the object, we would call `get_cache_shape`. When we want to obtain the maximum cache length, we would do something like `max_length = None if cache.is_sliding else cache.get_max_length()`

WDYT?

**Review by gante — COMMENTED** [review]
LGTM, thank you for working on this 💛  

One design question about sliding caches

File: src/transformers/models/mistral/modeling_mistral.py (hunk 1 above)
Comment (zucchini-nlp) [reply src/transformers/models/mistral/modeling_mistral.py:895]: Oh missed the moment we got rid of `max-length` in sliding window. Sounds good but I am not sure if there are cases when we want to know the `max_length` and not `cache_length`. Oke, let me go through the repo and check all possible cases :)

Comment (gante) [reply src/transformers/models/mistral/modeling_mistral.py:895]: > but I am not sure if there are cases when we want to know the max_length and not cache_length

I think the same :D but let's double-check 🙏

File: src/transformers/cache_utils.py [hunk 2]
```diff
@@ -63,15 +67,22 @@ def get_seq_length(self, layer_idx: Optional[int] = 0) -> int:
         raise NotImplementedError("Make sure to implement `get_seq_length` in a subclass.")
 
     def get_max_length(self) -> Optional[int]:
-        """Returns the maximum sequence length of the cached states, if there is any."""
+        """
+        Returns the maximum sequence length of the cache object. The method is deprecated,so use
+        `get_max_cache_shape` instead
+        """
         raise NotImplementedError("Make sure to implement `get_max_length` in a subclass.")
```
Comment (gante) [inline src/transformers/cache_utils.py:74]:
```suggestion
        <add the warning here>
        return self.get_max_cache_shape()
```

Suggestion: this as opposed to updating all cache classes :) on the cache classes, `def get_max_length(self) ...` would be removed

File: src/transformers/cache_utils.py [hunk 3]
```diff
@@ -62,16 +65,16 @@ def get_seq_length(self, layer_idx: Optional[int] = 0) -> int:
         # TODO: deprecate this function in favor of `cache_position`
         raise NotImplementedError("Make sure to implement `get_seq_length` in a subclass.")
 
-    def get_max_length(self) -> Optional[int]:
```
Comment (gante) [inline src/transformers/cache_utils.py:65]: We will need a deprecation cycle for `get_max_length` before we can delete it :P

(in the raised warning, make sure you mention that users should use `get_max_cache_shape` instead)

File: src/transformers/cache_utils.py [hunk 4]
```diff
@@ -28,6 +28,9 @@ class Cache(torch.nn.Module):
     Base, abstract class for all caches. The actual data structure is specific to each subclass.
     """
 
+    is_static = False
+    is_sliding = False
```
Comment (gante) [inline src/transformers/cache_utils.py:32]: (double-check whether this new attribute is correct, after rebasing. I think we've added new cache classes since your original commit. In the encoder-decoder cache, these attributes should be loaded from the decoder cache)

**Review by gante — APPROVED** [review]
LGTM, other than the deprecation cycle and the need to double-check attributes 🤗

File: src/transformers/cache_utils.py (hunk 4 above)
Comment (zucchini-nlp) [reply src/transformers/cache_utils.py:32]: yes, I ran tests for generation in llama and all compile-static tests in whisper. Rebased main and `fix-copies` propagated changes to some new models

File: src/transformers/cache_utils.py [hunk 5]
```diff
@@ -932,8 +944,8 @@ def get_seq_length(self, layer_idx: Optional[int] = 0) -> int:
             return 0
         return self.key_cache[layer_idx].shape[-2]
 
-    def get_max_length(self) -> Optional[int]:
-        """Returns the maximum sequence length of the cached states."""
+    def get_max_cache_shape(self) -> Optional[int]:
```
Comment (ArthurZucker) [inline src/transformers/cache_utils.py:947]: is there a specific reason for the name change ?

File: src/transformers/models/bloom/modeling_bloom.py [hunk 6]
```diff
@@ -799,10 +799,14 @@ def _update_causal_mask(
...
         # to infer the attention mask.
         past_seen_tokens = past_key_values.get_seq_length() if past_key_values is not None else 0
-        using_static_cache = isinstance(past_key_values, StaticCache)
 
         # When output attentions is True, sdpa implementation's forward method calls the eager implementation's forward
-        if self.config._attn_implementation == "sdpa" and not using_static_cache and not output_attentions:
+        if (
+            self.config._attn_implementation == "sdpa"
+            and past_key_values is not None
+            and not past_key_values.is_static
```
[the remark this answers is not in the document]
Comment (ArthurZucker) [reply src/transformers/models/bloom/modeling_bloom.py:812]: Aligned with my previous comment where we need a general `past_key_value.skip_sdpa_correction()` for example

File: src/transformers/models/bloom/modeling_bloom.py [hunk 7]
```diff
@@ -819,8 +823,8 @@ def _update_causal_mask(
         dtype, device = input_tensor.dtype, input_tensor.device
         min_dtype = torch.finfo(dtype).min
         sequence_length = input_tensor.shape[1]
-        if using_static_cache:
-            target_length = past_key_values.get_max_length()
+        if past_key_values is not None and past_key_values.is_static:
```
Comment (ArthurZucker) [inline src/transformers/models/bloom/modeling_bloom.py:826]: same here

**Review by ArthurZucker — COMMENTED** [review]
Thanks for the in depth changes! Not convinced we have to change the name of the function, and also convinced we should have abstraction on what we call. Checking for is static IMO is not a good way forward. It might be a good internal helper, but in the modeling we would prefer not having to check that wdyt? 

Also fine for this specific case TBH! Just not sure it's gonna scale

File: src/transformers/cache_utils.py (hunk 5 above)
Comment (zucchini-nlp) [reply src/transformers/cache_utils.py:947]: The only reason is to be more specific what we mean by `max_length` since I wanted to add `max_length` in sliding cache, which currently doesn't have any (returns `None`). And as comment in code explains it was done with idea that the cache object technically handles infinite amount of tokens, so no max length

But we never want to know how many tokens can cache technically handle, our checks are all about max capacity of particular cache instance

File: src/transformers/models/bloom/modeling_bloom.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (zucchini-nlp) [reply src/transformers/models/bloom/modeling_bloom.py:812]: Hmm, maybe we can make a better check to not use `isinstance()` on cache objects. We are having more varieties of cache now and it is mostly either dynamic or static type of some caching method, so it would be nice to not rely on `isinstance()` checks in general

We can remove modeling changes in this PR and use the new attributes in specific cases (linked comment from PR description). What I need from this PR currently is the `get_max_length` that doesn;t return `None` anymore

File: src/transformers/cache_utils.py (hunk 5 above)
Comment (ArthurZucker) [reply src/transformers/cache_utils.py:947]: Let's add this in the comment for example! I was not sure 🤗

File: src/transformers/models/bloom/modeling_bloom.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (ArthurZucker) [reply src/transformers/models/bloom/modeling_bloom.py:812]: Yeah Also want to avoid .is_static and isinstance etc. Not sure how to abstract best tho! 

Okay, let's remove then for now

[the remark this answers is not in the document]
Comment (zucchini-nlp) [reply src/transformers/models/bloom/modeling_bloom.py]: okey, done and rebased main

**Review by ArthurZucker — APPROVED** [review]
Thanks!
